### ISSUE TITLE
Problems with parameterized magnetic field

### ISSUE THREAD

**Issue by kmcdermo** [cerati/mkFit#124]
As we have seen a few times now, the efficiency for both building and backward fit degrades when the parameterized magnetic field is turned on.  It seems to be affecting the transition region the most, and not as noticeable elsewhere.

There have already been some ideas on how to check what is happening.  I wanted to post this here for quick feedback.  Please feel free to add insights as it comes.

- [ ] Check the value of magnetic field from what we expect to what is actually computed for each propagation.  

For reference, I am linking the delta for the magnetic field map between a static 3.8 T field and the parameterized one made from a simple macro (magfield.C): https://kmcdermo.web.cern.ch/kmcdermo/keep_old_mictrk/magfield/fieldmap/

**Comment by areinsvo** [comment]
Revisiting this issue:
[Current head of devel (after PR252)](http://areinsvo.web.cern.ch/areinsvo/MkFit/ChargeAssignment/April2020/baseline/SIMVAL_MTV_SEED) vs 
turning [on parametric b field](http://areinsvo.web.cern.ch/areinsvo/MkFit/ChargeAssignment/April2020/paramBField/SIMVAL_MTV_SEED) vs 
turning it on [only for interlayer propagation](http://areinsvo.web.cern.ch/areinsvo/MkFit/ChargeAssignment/April2020/paramBFieldInterOnly/SIMVAL_MTV_SEED) vs 
turning it on [only in the endcaps](http://areinsvo.web.cern.ch/areinsvo/MkFit/ChargeAssignment/April2020/paramBFieldEndcapOnly/SIMVAL_MTV_SEED). 

As you can see, the performance is slightly worse for the parametric b field, but the difference is not nearly as large as it was two years ago. This is definitely worth including in the list of improvements for the new tuning.

Head of devel (constant b field)
![baseline_eff](https://user-images.githubusercontent.com/7669806/81426590-0741fc00-911f-11ea-8473-25431ea280d1.png)

Parametric b field 
![paramBfield_eff](https://user-images.githubusercontent.com/7669806/81426580-027d4800-911f-11ea-94d9-ce251ff7ea0b.png)

For what it's worth, I did double check and compare our implementation to that [in CMSSW](https://cmssdt.cern.ch/lxr/source/MagneticField/ParametrizedEngine/src/ParabolicParametrizedMagneticField.cc) and everything seems consistent.

**Comment by areinsvo** [comment]
One more set of plots before I add this to Pandora's Box. I ran mkFit within CMSSW with the parametric B field turned on for inter-layer propagation only (orange) or inter-layer propagation and the backward fit (black). Nominal CMSSW (blue) and mkFit (red) are also shown. The full set of plots can be found [here](http://areinsvo.web.cern.ch/areinsvo/MkFit/CMSSW_Integration/PandorasBox/plots_BField). 

The effects are mostly within the noise, except for the pt residuals with respect to eta:
<img width="446" alt="Screen Shot 2020-05-20 at 3 27 47 PM" src="https://user-images.githubusercontent.com/7669806/82494394-1d50a480-9aaf-11ea-8569-332a06cb46c1.png">

The residuals in the endcaps get closer to CMSSW but further from 0. This is contradictory to what we see in our standalone plots, where the pt diffs improve (see below). The difference probably comes from the details of the CMSSW final fit, and it probably isn't worth investigating further now. Although it would be really useful to have plots of pt resolution vs eta in our standalone validation. 

Constant b field:
![SKL-SP_CMSSW_TTbar_PU50_bestmatch_dinvpt_build_pt0p0_SIMVALSEED](https://user-images.githubusercontent.com/7669806/82495216-63f2ce80-9ab0-11ea-91c2-4676ca3c8ced.png)

Parameterized b field used for interlayer propagation:
![SKL-SP_CMSSW_TTbar_PU50_bestmatch_dinvpt_build_pt0p0_SIMVALSEED-1](https://user-images.githubusercontent.com/7669806/82495381-a1eff280-9ab0-11ea-9034-db33ce62450b.png)

The timing is essentially identical with and without the parametric b field, at least when using a single thread within CMSSW.

**Comment by slava77** [comment]
> The residuals in the endcaps get closer to CMSSW but further from 0. This is contradictory to what we see in our standalone plots, where the pt diffs improve (see below).

I'm trying to understand how to match the two plots: one is deltaPt/ptSim (dimensionless), while the other is deltaPt/(ptSim*ptReco) (in units of 1/GeV). Wild-guessing an average pt of 2 GeV, I should be looking at the bias/mode "feature" of about 0.01/GeV in the mkFIt validation d(1/pt) histogram. 
For some more consistency it would be best to match the bins and variables so that we are not trying to compare tails in one case and core in the other and if these may differ.

**Comment by slava77** [comment]
>  For some more consistency it would be best to match the bins and variables so that we are not trying to compare tails in one case and core in the other and if these may differ.

oh, and there is an obvious eta dependence part: the concern in the comparison with the MTV is localized in the endcap, while the mkFit val plot is integrated for all eta

**Comment by areinsvo** [comment]
@slava77 I agree, these plots are not an apples-to-apples comparison, so maybe it's overstating it to say that the two sets of plots are contradictory. All I meant was that the **pt diffs get better** in our standalone validation, but the only noticeable effect in MTV is that the **pt residuals get worse**. As I mentioned earlier, it would be really useful to have plots of pt resolution vs eta in our standalone validation.
